domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/FirestoreDataController.kt

```diff
@@ -0,0 +1,169 @@
+    when (networkConnectionUtil.getCurrentConnectionStatus()) {
+      NetworkConnectionUtil.ProdConnectionStatus.NONE -> cacheEventForFirestore(eventLog)
+      else -> authenticateAndUploadToFirestore(eventLog)
+    }
+  }
+
+  private suspend fun authenticateAndUploadToFirestore(eventLog: EventLog) {
+    if (authenticationWrapper.getCurrentSignedInUser() == null) {
+      when (val signInResult = authenticationWrapper.signInAnonymously().await()) {
+        is AsyncResult.Success -> {
+          consoleLogger.i("FirestoreDataController", "Sign in succeeded")
+          eventLogger.uploadEvent(eventLog)
+        }
+        is AsyncResult.Failure -> {
+          consoleLogger.e(
+            "FirestoreDataController",
+            "Sign in failed with cause ${signInResult.error}"
+          )
+          cacheEventForFirestore(eventLog)
+        }
+        is AsyncResult.Pending -> {} // no-op
+      }
+    } else {
+      eventLogger.uploadEvent(eventLog)
+    }
+  }
+
+  /**
+   * Adds an event to the storage.
+   *
+   * The [eventLog] is added to the store if the size of the store isn't exceeding
+   * [logStorageCacheSize]. If the limit is exceeded then the least recent event is removed from the
+   * [firestoreEventsStore].
+   */
+  private fun cacheEventForFirestore(eventLog: EventLog) {
+    firestoreEventsStore.storeDataAsync(updateInMemoryCache = true) { eventLogs ->
+      val storeSize = eventLogs.eventLogsToUploadList.size
+      if (storeSize + 1 > logStorageCacheSize) {
+        val eventLogRemovalIndex = getLeastRecentEventIndex(eventLogs)
+        if (eventLogRemovalIndex != null) {
+          return@storeDataAsync eventLogs.toBuilder()
+            .removeEventLogsToUpload(eventLogRemovalIndex)
+            .addEventLogsToUpload(eventLog)
+            .build()
+        } else {
+          val exception =
+            IllegalStateException(
+              "Least Recent Event index absent -- FirestoreLogStorageCacheSize is 0"
+            )
+          consoleLogger.e("FirestoreDataController", "Failure while caching event.", exception)
+          exceptionLogger.logException(exception)
+        }
+      }
+      return@storeDataAsync eventLogs.toBuilder().addEventLogsToUpload(eventLog).build()
+    }.invokeOnCompletion {
+      it?.let { consoleLogger.e("FirestoreDataController", "Failed to store event log.", it) }
+    }
+  }
+
+  /**
+   * Returns the index of the least recent event from the existing store on the basis of recency and
+   * priority.
+   */
+  private fun getLeastRecentEventIndex(oppiaEventLogs: OppiaEventLogs): Int? =
+    oppiaEventLogs.eventLogsToUploadList.withIndex().minByOrNull { it.value.timestamp }?.index
+
+  /** Returns a data provider for log reports that have been recorded for upload. */
+  fun getEventLogStore(): DataProvider<OppiaEventLogs> = firestoreEventsStore
+
+  /** Removes the first log report that had been recorded for upload. */
+  fun removeFirstEventLogFromStore() {
```

Comment:
I originally had this as the function was called in loguploadworker to remove logs after they have been uploaded. I have made a change so that it is called after each log has been uploaded within this controller. WDYT about this?

Reply:
Yes, I changed access to private